ocfweb/docs/docs/staff/procedures/deploying-services.md

```diff
@@ -0,0 +1,373 @@
+[[!meta title="Deploying Services"]]
+
+## Overview
+Are you looking to deploy a new service to the OCF Kubernetes cluster? This
+document will cover all of the steps required to do so. Do note that this is **not** a
+substitute for a Kubernetes tutorial or a Docker tutorial _(there are many
+resources online for that)_ but a guide for getting your service running on the
+OCF. There are a lot of pieces of OCF infrastructure at play here, but this
+should serve as a gentle introduction into how things are done.
+
+This `HOWTO` uses one of the OCF's simplest services as an example:
+[Templates][templates]. Templates is service used internally by OCF staff
+serving 'copy-pasteable' email templates. You can use [[git|doc
+staff/backend/git]] to `clone` this [repo](https://github.com/ocf/templates/blob/master/kubernetes/templates.yml.erb)
+, or just follow along looking at the files. If you want to blaze ahead,
+this repo should serve as a good example of the required components in the
+repo for deployment.
+
+## Prerequisites
+If you're looking to deploy your service, you should first have a service to
+deploy. This is normally done with a repository underneath the
+[OCF's Github organization](https://github.com/ocf). If you have the correct
+permissions, you should be able to transfer, fork, or just create a
+repository in the org.
+
+### Docker & Jenkins
+
+For your service to be deployed onto the Kubernetes cluster, it must have
+a `Dockerfile` so that we can containerize the service. Nothing OCF specific
+here, just include whatever your service needs.
+
+[[Jenkins|doc staff/backend/jenkins]] will try to build that container,
+and then send the image to the OCF Docker server. This has to be specified in
+the `Makefile` at the root of the repository. Here's part of what templates has:
+
+```
+BIN := venv/bin
+
+DOCKER_REVISION ?= testing-$(USER)
+DOCKER_TAG = docker-push.ocf.berkeley.edu/templates:$(DOCKER_REVISION)
+RANDOM_PORT := $(shell expr $$(( 8000 + (`id -u` % 1000) + 1 )))
+
+
+.PHONY: dev
+dev: cook-image
+    @echo "Will be accessible at http://$(shell hostname -f ):$(RANDOM_PORT)/"
+    docker run --rm -p "$(RANDOM_PORT):8000" "$(DOCKER_TAG)"
+
+.PHONY: cook-image
+cook-image:
+    docker build --pull -t $(DOCKER_TAG) .
+
+.PHONY: push-image
+push-image:
+    docker push $(DOCKER_TAG)
+```
+
+The important targets to pay attention here are `cook-image` and `push-image`,
+Jenkins will try to run these as part of the deployment pipeline that continuously
+deploys to Kubernetes.
+
+Finally, we have to include a `Jenkinsfile` to the repository, so that Jenkins
+knows how we want it to go about deploying the service. In this case, it's just
+specifying that we want it to go through the pipeline.
+
+```
+servicePipeline(
+    upstreamProjects: ['ocf/dockers/master'],
+)
+```
```

Comment:
This part is a little confusing to me, as in I described my understanding of it, but in other projects an empty `upstreamProjects` array also seemed to work fine. Perhaps @jvperrin could chime in on this?

Reply:
Yeah, that's a great question! The only thing this does is mean that a service build pipeline gets kicked off if the upstream one listed here succeeds. The reason why you see `ocf/dockers/master` in a number of services is when they are both (1) built off the OCF base docker images and (2) redeploying them can be done smoothly. These redeploys are mainly done so that security updates are pulled into service containers in a timely manner, so it doesn't have to be a daily thing but could probably be weekly or something instead.

For instance, [ocfweb has both ocflib and dockers](https://github.com/ocf/ocfweb/blob/303e0671ace54ce6e70114d7f36908c329d41a31/Jenkinsfile#L2) as upstreams because it depends on the latest versions of both of those repos, and because redeploying after either one finishes building successfully can be done smoothly. [slackbridge](https://github.com/ocf/slackbridge/blob/9dd3ac36b9e62ebb9b886435f6a2fe6458406431/Jenkinsfile#L2) on the other hand doesn't have any upstream projects listed despite being built off the OCF base image because every time it rebuilt it causes large join/quit spam on IRC that wasn't a pleasant experience.

Essentially, to figure out what's listed here for a new service, you only need to know what base image the service is using, and how disruptive a deploy would be.